dwave/system/testing.py

```diff
@@ -31,30 +31,101 @@
+        G = dnx.chimera_graph(*shape,
+                              node_list=nodelist,
+                              edge_list=edgelist)
+        
+    elif topology_type == 'pegasus':
+        if len(shape) != 1:
+            raise ValueError('shape is incompatible with a pegasus lattice.')
+        G = dnx.pegasus_graph(shape[0],
+                              node_list=nodelist,
+                              edge_list=edgelist)
+
+    elif topology_type == 'zephyr':
+        if len(shape) not in (1, 2):
+            raise ValueError('shape is incompatible with a zephyr lattice.')
+        G = dnx.zephyr_graph(*shape,
+                             node_list=nodelist,
+                             edge_list=edgelist)
+    else:
+        # Alternative could be to create a standard network graph and
+        # issue a warning. Requires new dependency on networkx.
+        raise ValueError('topology_type does not match a known'
+                         'QPU architecure')
+    return G
     
+class MockDWaveSampler(dimod.Sampler, dimod.Structured):
+    """Mock sampler modeled after DWaveSampler that can be used for tests.
+
+    Properties and topology parameters are populated qualitatively matching
+    online systems, and a placeholder sampler routine based on steepest descent
+    is instantiated.
+
     Args:
+        nodelist (iterable of ints, optional):
+            List of nodes to include. When not specified, all nodes
+            compatible with the default topology type and shape are
+            added.
+
+        edgelist (iterable of (int,int) tuples, optional):
+            List of edges to include. When not specified, all edges
+            compatible with the default topology type and shape are
+            added.
+
+        properties (dictionary, optional):
+            A dictionary of properties, to update default values.
+
         broken_nodes (iterable of ints, optional):
-            List of nodes to exclude (along with associated edges). For 
-            emulation of unyielded qubits.
+            List of nodes to exclude (along with associated edges), for
+            emulation of unyielded qubits. This parameter is made redundant
+            by the use of nodelist.
         
         broken_edges (iterable of (int,int) tuples, optional):
-            List of edges to exclude. For emulation of unyielded edges.
-        
+            List of edges to exclude, for emulation of unyielded edges.
+            This parameter is made redundant by the use of edgelist.
+
         topology_type (string, optional, default='chimera'):
-            QPU topology being emulated. Note that for Pegasus emulation the 
-            fabric_only=True graph is presented. Supported options are
-            'chimera', 'pegasus' or 'zephyr'
+            If 'topology' is in ``properties``, this argument is ignored
+            in favour of ``properties``['topology']['type']. Otherwise
```

Comment:
In typical case properties might be taken directly from a qpu: DWaveSampler().properties. These properties contains a variety of interdependent fields (e.g topology, num_qubits, qubits, etc). Allowing the user to override the topology type only is I think a very bad idea. An expert user who understands this can simply delete the topology field or overwrite it, and the default protects naive users. happy to be over ruled if your really insist.

Reply:
Ok, I'll have to think about this. Currently, it looks quite square to me.

Reply:
I think, after some thought, just throw an error if the arguments contradict one another is safest, having contradictory arguments would never be a sensible use case that I can think of.

Reply:
The more I look at this interface, the more confusing it looks.

- `properties` include both `topology` and node/edge list, and specifying `properties` will override both, but keep `nodelist`/`edgelist` attributes (out of sync)
- not clear how `nodelist`/`edgelist` interact with topology -- only one is needed. In fact, networkx `solver_graph` generation is needed only when just topology is specified.
- but topology can be specified in properties as well

Erroring on contradicting arguments is sensible, but not sure it's possible or straightforward to implement (the `properties` + something else case).

Progressive update/override seems clearer:
- hard-coded defaults are overridden by user-supplied properties
- such updated properties are then updated by nodelist/edgelist and topology
- nodelist/edgelist attributes are generated based off properties above
- nodelist/edgelist are updated from broken_*

In addition, we could add these factories:
- `from_topology(type, shape)`
- `from_graph(nodelist, edgelist)`

Reply:
I'd be happy to see this full solution implemented, but I did something simpler that I think is still fit for purpose. Contradictions between properties and nodelist for example are sill possible.